File: CCFF_2024_AER--M.R
################################################################################
# Replication "On Binscatter"
# Cattaneo, Crump, Farrell and Feng (2023)
# Date: 8-JUL-2023
# Moretti (2021) Application
################################################################################

# source("CCFF_2023_Binscatter_M.R")

rm(list=ls())

library(haven)
library(binsreg)
library(ggplot2)
library(lfe)

data <- read_dta(file='CCFF_2023_M_1.dta')
data <- subset(data, select=c(y,x, bea_code,year, zd2))
data = data[complete.cases(data), ]

######################################################################################
# (6) Relation between productivity of top inventors and high-tech clusters
######################################################################################

################################################################################
# 6(a)
################################################################################
png('graphs/M_scatter.png')
plot(data$x, data$y, xlab="log Cluster Size", ylab="log Number of Patents per Inventor per Year", pch=19, col='blue')
dev.off()

################################################################################
# 6(b)
################################################################################
reg <- felm(y ~ as.factor(bea_code) + as.factor(year) + as.factor(zd2), data)
resid_y <- resid(reg) + mean(data$y)
reg <- felm(x ~ as.factor(bea_code) + as.factor(year) + as.factor(zd2), data)
resid_x <- resid(reg) + mean(data$x)

res <- binsreg(resid_y, resid_x, nbins=40,  polyreg=1)
dev.off()
tmp <- res$bins_plot + xlab("log Cluster Size") + ylab("log Number of Patents per Inventor per Year") + ylim(-.35, -.15) + xlim(-5.5, -2.5)
png('graphs/M_binscatter.png')
plot(tmp)
dev.off()

################################################################################
# 6(c)
################################################################################
resid_inc <-res$data.plot$`Group Full Sample`$data.dots

res <- binsreg(resid_y, resid_x, nbins=40,  line=c(0,0), linegrid=1000)
dev.off()
line <- res$data.plot$`Group Full Sample`$data.line
tmp <- res$bins_plot + geom_line(data=line[!is.na(line$fit),], aes(x=x,y=fit, color='orange')) + xlab("log Cluster Size") + ylab("log Number of Patents per Inventor per Year") + ylim(-.35, -.15) + xlim(-5.5, -2.5)
png('graphs/M_binscatter_pwc.png')
plot(tmp)
dev.off()

################################################################################
# 6(d)
################################################################################

#### Note that Figure 6(d) may differ from the manuscript because of the evaluation point of the control variables; see the discussion in the Supplemental Appendix

w <- model.matrix(~ as.factor(bea_code) + as.factor(year) + as.factor(zd2) - 1, data=data)

res <- binsreg(data$y, data$x, w, randcut=1, nbins=40)
tmp <- res$bins_plot + geom_point(data=resid_inc, aes(x=x, y=fit, color='orange'))+ xlab("log Cluster Size") + ylab("log Number of Patents per Inventor per Year") + ylim(-.4,0) + xlim(-8,-1)
dev.off()
png('graphs/M_covariateAdjustments_binscatter.png')
plot(tmp)
dev.off()


################################################################################
# 6(e)
################################################################################

#### Note that Figure 6(e) may differ from the manuscript because of the evaluation point of the control variables; see the discussion in the Supplemental Appendix

data <- read_dta(file='CCFF_2023_M_2.dta')
data <- subset(data, select=c(y, x, bea_code, year, zd2, cluster1))
data2 <- data[complete.cases(data), ]

w <- model.matrix(~ as.factor(bea_code) + as.factor(year) + as.factor(zd2) - 1, data2)

res <- binsreg(data2$y, data2$x, w, cluster=data2$cluster1, randcut=1, cb=T)
tmp <- res$bins_plot + xlab("log Cluster Size") + ylab("log Number of Patents per Inventor per Year") + ylim(-.5,.1) + xlim(-8,-1)
dev.off()
png('graphs/M_confidenceBand.png')
plot(tmp)
dev.off()

#### Note that Figure 6(f) is omitted because of the large sample size; please see the Stata code for replication

File: CCFF_2024_AER--M.py
################################################################################
# Replication "On Binscatter"
# Cattaneo, Crump, Farrell and Feng (2024)
# Date: 05-JAN-2024
# M Application
################################################################################
import pandas as pd
import numpy as np
from binsreg import *
import sklearn
import matplotlib.pyplot as plt
import statsmodels.api as sm 
from sklearn_pandas import DataFrameMapper

data_og = pd.io.stata.read_stata('../data/CCFF_2023_M_1.dta')
data_og['constant'] = 1

################################################################################
# Figure 6(a) Scatter plot
################################################################################
plt.scatter(data_og['x'], data_og['y'])
plt.xlabel("log Cluster Size")
plt.ylabel("log Number of Patents per Inventor per Year")

################################################################################
# Figure 6(b)-6(d) Scatter plot
################################################################################
data = data_og
data = data[['constant','bea_code','zd2','year','y','x']].dropna()

cov = data[['constant']].to_numpy()
bea_flags = DataFrameMapper([('bea_code', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
zd2_flags = DataFrameMapper([('zd2', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
year_flags = DataFrameMapper([('year', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
X = np.concatenate((cov, bea_flags, year_flags, zd2_flags),axis=1)

reg = sm.regression.linear_model.OLS(data['y'], X).fit()
resid_y = reg.resid + np.mean(data['y'])
reg = sm.regression.linear_model.OLS(data['x'], X).fit()
resid_x = reg.resid + np.mean(data['x'])

res = binsreg(resid_y,resid_x,nbins=40, polyreg=1, plotxrange=[-5.5,-2.5], plotyrange=[-.35,-.15])

x = list(range(-5500,-2500))
x = pd.DataFrame([v/1000 for v in x], columns=['x'])
x['dots'] = np.zeros(len(x))

dots = res.data_plot[0].dots
bins = res.data_plot[0].data_bin
poly = res.data_plot[0].poly
bin_last = bins['bin_id'].iloc[-1]

bin_counter = 0
bin_right = float(bins[bins['bin_id']==bin_counter+1]['right.endpoint'].iloc[0]) # discrep
bin_dot = float(dots[dots['bin']==bin_counter].fit.iloc[0])
for index, row in x.iterrows():
    if row['x'] > bin_right:
        bin_counter += 1
        if bin_counter < bin_last:
            bin_right = float(bins[bins['bin_id']==(bin_counter+1)]['right.endpoint'].iloc[0])
            bin_dot = float(dots[dots['bin']==bin_counter].fit.iloc[0])
    
    x.iloc[index, 1] = bin_dot

plt.plot(x['x'],x['dots'])
plt.scatter(dots['x'], dots['fit'], s=10, c="blue")
plt.plot(poly['x'],poly['fit'])
plt.xlim([-5.5,-2.5])
plt.ylim([-.35,-.15])

dots2 = res.data_plot[0].dots
res = binsreg(data['y'], data['x'], X, nbins=40, noplot=True)
dots1 = res.data_plot[0].dots
plt.scatter(dots1['x'],dots1['fit'])
plt.scatter(dots2['x'],dots2['fit'])

################################################################################
# Figure 6(e) Optimal Choice of J/Confidence Band
################################################################################
data_og = pd.io.stata.read_stata('../data/CCFF_2023_M_2.dta')
data_og['constant'] = 1

data = data_og[['constant','bea_code','zd2','year','y','x','cluster1']].dropna()

cov = data[['constant']].to_numpy()
bea_flags = DataFrameMapper([('bea_code', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
zd2_flags = DataFrameMapper([('zd2', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
year_flags = DataFrameMapper([('year', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
X = np.concatenate((cov, bea_flags, year_flags, zd2_flags),axis=1)

res = binsreg(data['y'], data['x'], X, cluster=data['cluster1'])

################################################################################
# Figure 6(f) Optimal Choice of J/Confidence Band (full specification)
################################################################################

data = data_og[['constant','class','cluster_bea_class','cluster1','cluster_class_year','cluster_zd_year','inventor_id','cluster_bea_year','org_new','y','x']].dropna()

c1_flags = DataFrameMapper([('class', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
c2_flags = DataFrameMapper([('cluster_bea_class', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
c3_flags = DataFrameMapper([('cluster1', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
c4_flags = DataFrameMapper([('cluster_class_year', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
c5_flags = DataFrameMapper([('cluster_zd_year', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
c6_flags = DataFrameMapper([('inventor_id', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
c7_flags = DataFrameMapper([('cluster_bea_year', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]
c8_flags = DataFrameMapper([('org_new', sklearn.preprocessing.LabelBinarizer())]).fit_transform(data)[:,1:]

X = np.concatenate((X,c1_flags,c2_flags,c3_flags,c4_flags,c5_flags,c6_flags,c7_flags,c8_flags),axis=1)

res = binsreg(data['y'], data['x'], X, cluster=data['cluster1'], randcut=1, cb=(1,1))